File: 5.binarySearch/367.ValidPerfectSquare/main.py
from typing import List

# https://leetcode.com/problems/valid-perfect-square/

# TC: O(log(num)), SC: O(1)
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        l, r = 1, num
        while l <= r:
            m = l + (r - l)//2
            square = m*m
            if square == num:
                return True
            if square > num:
                r = m - 1
            else:
                l = m + 1
        return False

File: 5.binarySearch/367.ValidPerfectSquare/main.go
package main

// https://leetcode.com/problems/valid-perfect-square/
func isPerfectSquare(num int) bool {
	low := 1
	high := num

	for low <= high {
		mid := low + (high-low)/2
		square := mid * mid
		if square == num {
			return true
		}
		if square > num {
			high = mid - 1
		} else {
			low = mid + 1
		}
	}

	return false
}
